packages/muelu/src/Graph/StructuredAggregation/MueLu_AggregationStructuredAlgorithm_def.hpp

```diff
@@ -134,11 +138,184 @@ namespace MueLu {
       procWinner[nodeIdx]   = ghostedCoarseNodeCoarsePIDs[ghostedCoarseNodeCoarseLID];
       aggStat[nodeIdx]      = AGGREGATED;
       --numNonAggregatedNodes;
+
+    } // Loop over fine points
+
+  }
+
+
+  template <class LocalOrdinal, class GlobalOrdinal, class Node>
+  void AggregationStructuredAlgorithm<LocalOrdinal, GlobalOrdinal, Node>::
+  BuildAggregates(const ParameterList& params, const GraphBase& graph, Aggregates& aggregates, RCP<CrsGraph>& myGraph) const {
+    Monitor m(*this, "BuildAggregates");
+
+    int myRank = graph.GetDomainMap()->getComm()->getRank();//
+
+    RCP<Teuchos::FancyOStream> out;
+    if(const char* dbg = std::getenv("MUELU_STRUCTUREDALGORITHM_DEBUG")) {
+      out = Teuchos::fancyOStream(Teuchos::rcpFromRef(std::cout));
+      out->setShowAllFrontMatter(false).setShowProcRank(true);
+    } else {
+      out = Teuchos::getFancyOStream(rcp(new Teuchos::oblackholestream()));
+    }
+
+    std::string coupling = params.get<std::string>("aggregation: coupling");
+    const bool coupled = (coupling == "coupled" ? true : false);
+    RCP<IndexManager> geoData = aggregates.GetIndexManager();
+    Array<LO>  ghostedCoarseNodeCoarseLIDs;
+    Array<int> ghostedCoarseNodeCoarsePIDs;
+    Array<GO> ghostedCoarseNodeCoarseGIDs;
+
+    *out << "Extract data for ghosted nodes" << std::endl;
+    geoData->getGhostedNodesData(graph.GetDomainMap(), ghostedCoarseNodeCoarseLIDs,
+                                 ghostedCoarseNodeCoarsePIDs, ghostedCoarseNodeCoarseGIDs);
+
+    LO iGhosted, jGhosted, kGhosted, iCoarse, jCoarse, kCoarse, iRem, jRem, kRem;
+    LO ghostedCoarseNodeCoarseLID, aggId, rate;
+    LO dim = geoData->getNumDimensions();
+    LO rowCount = 0;
+    GO myGID;
+    bool iIsCoarse, jIsCoarse, kIsCoarse;
+    Array<LO> offsets = geoData->getOffsets();
+    Array<LO> colIndex(geoData->getNumLocalFineNodes()*std::pow(2,dim));
+    Array<LO> rowPtr(geoData->getNumLocalFineNodes()+1);
+    rowPtr[0] = 0;
+    ArrayRCP<size_t> nnzOnRow(geoData->getNumLocalFineNodes());
+
+    *out << "Loop over fine nodes and assign them to an aggregate and a rank" << std::endl;
+    for(LO nodeIdx = 0; nodeIdx < geoData->getNumLocalFineNodes(); ++nodeIdx) {
+      // Compute coarse ID associated with fine LID
+      geoData->getFineNodeGhostedTuple(nodeIdx, iGhosted, jGhosted, kGhosted);
```

Comment:
modify `getFineNodeGhostedTuple()` so that it returns an array `ijkGhosted` instead of three variables and then rewrite the code below as a loop